torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -449,6 +476,101 @@ void ProcessGroupAgent::listenLoop() {
   }
 }
 
+void ProcessGroupAgent::pollTimedOutRPCs() {
+  // Sleep until a future is available, or there are no futures and we are
+  // shutting down.
+  {
+    std::unique_lock<std::mutex> lock(futureTimeoutMutex_);
+    futureTimeoutCV_.wait(lock);
+  }
+
+  while (!shutdown_.load()) {
+    std::chrono::milliseconds sleepTime;
+    {
+      std::lock_guard<std::mutex> lock{futureMutex_};
+      // Estimate amount of time the first future will time out in, and sleep
+      // for that long.
+      if (futureTimeouts_.empty()) {
+        continue;
+      }
+      auto runningTime =
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+              std::chrono::steady_clock::now().time_since_epoch()) -
+          futureTimeouts_.begin()->first;
+      // TODO: will sleep for 0 ms when rpcTimeout_ is 0 (meaning infinite),
+      // which will result in busy polling for the lock If rpcTimeout_ is 0,
+      // should we have a default sleep interval here?
+      sleepTime =
+          std::max(rpcTimeout_ - runningTime, std::chrono::milliseconds(0));
```

Comment:
Right now, if the `futureTimeouts_` map is empty, or `rpcTimeout_` is set to 0, we basically won't sleep and simply go back to the beginning of the loop. This could result in a lot of lock contention, reducing efficiency. Should we have some default sleep time for this thread due to this?

Reply:
Thanks for the suggestion, this makes the code a lot simpler.